File: mudu_api/csharp/Mudu.cs
#nullable enable

global using UniCommandArgv = Universal.UniCommandArgv;
global using UniCommandResult = Universal.UniCommandResult;
global using UniCommandReturn = Universal.UniCommandReturn;
global using UniCommandReturnErr = Universal.UniCommandReturnErr;
global using UniCommandReturnKind = Universal.UniCommandReturnKind;
global using UniCommandReturnOk = Universal.UniCommandReturnOk;
global using UniDatType = Universal.UniDatType;
global using UniDatValue = Universal.UniDatValue;
global using UniError = Universal.UniError;
global using UniMessage = Universal.UniMessage;
global using UniOid = Universal.UniOid;
global using UniPrimitive = Universal.UniPrimitive;
global using UniPrimitiveValue = Universal.UniPrimitiveValue;
global using UniQueryArgv = Universal.UniQueryArgv;
global using UniQueryResult = Universal.UniQueryResult;
global using UniQueryReturn = Universal.UniQueryReturn;
global using UniQueryReturnErr = Universal.UniQueryReturnErr;
global using UniQueryReturnKind = Universal.UniQueryReturnKind;
global using UniQueryReturnOk = Universal.UniQueryReturnOk;
global using UniRecordField = Universal.UniRecordField;
global using UniRecordType = Universal.UniRecordType;
global using UniResultSet = Universal.UniResultSet;
global using UniSqlParam = Universal.UniSqlParam;
global using UniSqlStmt = Universal.UniSqlStmt;
global using UniTupleRow = Universal.UniTupleRow;

using MessagePack;

namespace Mudu.Api;

public static class Mudu
{
    public static CommandResponse Command(UniCommandArgv argv, MessagePackSerializerOptions? options = null)
    {
        return new CommandResponse(MuduSysCallApi.SysCommand(argv, options));
    }

    public static QueryResponse Query(UniQueryArgv argv, MessagePackSerializerOptions? options = null)
    {
        return new QueryResponse(MuduSysCallApi.SysQuery(argv, options));
    }

    public static byte[] Serialize(UniCommandArgv argv, MessagePackSerializerOptions? options = null)
    {
        return MuduSysCallApi.SerializeCommand(argv, options);
    }

    public static byte[] Serialize(UniQueryArgv argv, MessagePackSerializerOptions? options = null)
    {
        return MuduSysCallApi.SerializeQuery(argv, options);
    }

    public static UniCommandReturn DeserializeCommand(byte[] bytes, MessagePackSerializerOptions? options = null)
    {
        return MuduSysCallApi.DeserializeCommandResult(bytes, options);
    }

    public static UniQueryReturn DeserializeQuery(byte[] bytes, MessagePackSerializerOptions? options = null)
    {
        return MuduSysCallApi.DeserializeQueryResult(bytes, options);
    }
}

public readonly struct CommandResponse
{
    private readonly UniCommandReturn inner;

    public CommandResponse(UniCommandReturn inner)
    {
        this.inner = inner;
    }

    public UniCommandReturn Raw => inner;

    public bool IsOk => inner.Kind() == UniCommandReturnKind.Ok;

    public bool IsErr => inner.Kind() == UniCommandReturnKind.Err;

    public UniCommandResult? Result => IsOk ? UniCommandReturnOk.AsOk(inner).Inner : null;

    public UniError? Error => IsErr ? UniCommandReturnErr.AsErr(inner).Inner : null;

    public ulong? AffectedRows => Result?.AffectedRows;

    public UniCommandResult RequireOk()
    {
        if (IsOk)
        {
            return UniCommandReturnOk.AsOk(inner).Inner;
        }

        throw new global::System.InvalidOperationException(UniCommandReturnErr.AsErr(inner).Inner.ErrMsg);
    }
}

public readonly struct QueryResponse
{
    private readonly UniQueryReturn inner;

    public QueryResponse(UniQueryReturn inner)
    {
        this.inner = inner;
    }

    public UniQueryReturn Raw => inner;

    public bool IsOk => inner.Kind() == UniQueryReturnKind.Ok;

    public bool IsErr => inner.Kind() == UniQueryReturnKind.Err;

    public UniQueryResult? Result => IsOk ? UniQueryReturnOk.AsOk(inner).Inner : null;

    public UniError? Error => IsErr ? UniQueryReturnErr.AsErr(inner).Inner : null;

    public UniRecordType? TupleDesc => Result?.TupleDesc;

    public UniResultSet? ResultSet => Result?.ResultSet;

    public UniQueryResult RequireOk()
    {
        if (IsOk)
        {
            return UniQueryReturnOk.AsOk(inner).Inner;
        }

        throw new global::System.InvalidOperationException(UniQueryReturnErr.AsErr(inner).Inner.ErrMsg);
    }
}

File: mudu_api/rust/src/mudu.rs
use crate::error::ApiError;
use crate::mudu_sys;
use crate::types::{UniCommandResult, UniCommandReturn, UniQueryReturn};
use crate::{UniCommandArgv, UniError, UniQueryArgv, UniQueryResult, UniRecordType, UniResultSet};

pub struct Mudu;

impl Mudu {
    pub async fn command(argv: &UniCommandArgv) -> Result<CommandResponse, ApiError> {
        Ok(CommandResponse::new(mudu_sys::sys_command(argv).await?))
    }

    pub async fn query(argv: &UniQueryArgv) -> Result<QueryResponse, ApiError> {
        Ok(QueryResponse::new(mudu_sys::sys_query(argv).await?))
    }

    pub fn serialize_command(argv: &UniCommandArgv) -> Result<Vec<u8>, ApiError> {
        mudu_sys::serialize_command(argv)
    }

    pub fn serialize_query(argv: &UniQueryArgv) -> Result<Vec<u8>, ApiError> {
        mudu_sys::serialize_query(argv)
    }

    pub fn deserialize_command(bytes: &[u8]) -> Result<UniCommandReturn, ApiError> {
        mudu_sys::deserialize_command_result(bytes)
    }

    pub fn deserialize_query(bytes: &[u8]) -> Result<UniQueryReturn, ApiError> {
        mudu_sys::deserialize_query_result(bytes)
    }
}

#[derive(Debug, Clone)]
pub struct CommandResponse {
    inner: UniCommandReturn,
}

impl CommandResponse {
    pub fn new(inner: UniCommandReturn) -> Self {
        Self { inner }
    }

    pub fn raw(&self) -> &UniCommandReturn {
        &self.inner
    }

    pub fn is_ok(&self) -> bool {
        matches!(self.inner, UniCommandReturn::Ok(_))
    }

    pub fn is_err(&self) -> bool {
        matches!(self.inner, UniCommandReturn::Err(_))
    }

    pub fn result(&self) -> Option<&UniCommandResult> {
        match &self.inner {
            UniCommandReturn::Ok(result) => Some(result),
            UniCommandReturn::Err(_) => None,
        }
    }

    pub fn error(&self) -> Option<&UniError> {
        match &self.inner {
            UniCommandReturn::Ok(_) => None,
            UniCommandReturn::Err(error) => Some(error),
        }
    }

    pub fn affected_rows(&self) -> Option<u64> {
        self.result().map(|result| result.affected_rows)
    }

    pub fn require_ok(self) -> Result<UniCommandResult, UniError> {
        match self.inner {
            UniCommandReturn::Ok(result) => Ok(result),
            UniCommandReturn::Err(error) => Err(error),
        }
    }
}

#[derive(Debug, Clone)]
pub struct QueryResponse {
    inner: UniQueryReturn,
}

impl QueryResponse {
    pub fn new(inner: UniQueryReturn) -> Self {
        Self { inner }
    }

    pub fn raw(&self) -> &UniQueryReturn {
        &self.inner
    }

    pub fn is_ok(&self) -> bool {
        matches!(self.inner, UniQueryReturn::Ok(_))
    }

    pub fn is_err(&self) -> bool {
        matches!(self.inner, UniQueryReturn::Err(_))
    }

    pub fn result(&self) -> Option<&UniQueryResult> {
        match &self.inner {
            UniQueryReturn::Ok(result) => Some(result),
            UniQueryReturn::Err(_) => None,
        }
    }

    pub fn error(&self) -> Option<&UniError> {
        match &self.inner {
            UniQueryReturn::Ok(_) => None,
            UniQueryReturn::Err(error) => Some(error),
        }
    }

    pub fn tuple_desc(&self) -> Option<&UniRecordType> {
        self.result().map(|result| &result.tuple_desc)
    }

    pub fn result_set(&self) -> Option<&UniResultSet> {
        self.result().map(|result| &result.result_set)
    }

    pub fn require_ok(self) -> Result<UniQueryResult, UniError> {
        match self.inner {
            UniQueryReturn::Ok(result) => Ok(result),
            UniQueryReturn::Err(error) => Err(error),
        }
    }
}
